### ISSUE TITLE
Discrepancy between Centroidal Dynamics derivations

### ISSUE THREAD

**Issue by Michiel283** [stack-of-tasks/pinocchio#2093]
First of all, I would like to thank you for developing the Pinocchio library as it has helped me a lot during my research. I am currently trying to verify my Centroidal Dynamics model that I derived analytically, by comparing my result with the result obtained from Pinocchio. I expected the results to be identical, but this doesn't seem to be the case, and I was hoping someone could help me.

I derived an analytical expression for the CMM matrix using the David E. Orin's 2008 paper "_Centroidal Momentum Matrix of a humanoid robot: Structure and properties_". When I check my CMM matrix with the CMM matrix computed by ccrba algorithm, the two matrices are identical. Both matrices have the following block structure:
![Screenshot 2023-11-22 at 12-40-15 Thesis Project Kockelkoren](https://github.com/stack-of-tasks/pinocchio/assets/79510348/d1d144e9-3449-431a-b48a-c4bf352cee95)

However, when I compute dAg, you would expect that the time-derivative of Ag would have a similar block-structure as is also the case in my analytical derivation. When I compute the dAg matrix using dccrba or the computeCentroidalMapTimeVariation function, only the rows belonging to the linear momentum are identical. Moreover, the block-matrix which intuitively would be filled with zero's is nonzero in the dAg matrix computed by dccrba. 

When deriving the centroidal momentum hg, I derived the following analytic expression:
![Screenshot 2023-11-22 at 13-27-50 Thesis Project Kockelkoren](https://github.com/stack-of-tasks/pinocchio/assets/79510348/9a14ea99-3d29-4d56-8545-f390b520658e)

where A_B is the CMM matrix that maps the generalized joint rates to the momentum in the **body** frame, and X_G transforms this CMM matrix to the Centroidal frame G. Now, when computing the time-derivative of the centroidal momentum dhg, and thus also the time-derivative of the CMM, you would expect the following expression for dAg:
![Screenshot 2023-11-22 at 13-31-56 Thesis Project Kockelkoren](https://github.com/stack-of-tasks/pinocchio/assets/79510348/45325ac1-5149-4f30-b7e7-5a215f4d8ee5)

However, when looking at pinocchio computation, the Ag and the dAg seems to be computed in a different way. From my understanding, the CMM matrices are first computed for the the World frame and then transformed to the COM frame. This makes sense for the Ag matrix. However, by taking the time-derivative of the Ag matrix in the World frame, the translational transformation is neglected when taking the time-derivative of Ag. To my understanding, this mismatch would explain the non-zero values in the dAg matrix.
![Screenshot 2023-11-22 at 13-18-08 Thesis Project Kockelkoren](https://github.com/stack-of-tasks/pinocchio/assets/79510348/0d0cd790-4d35-4232-968d-23b6909473b2)

I hope you can help me to better understand how Pinocchio computes the centroidal dynamics. Thank you very much in advance.

Kind regards,
Michiel Kockelkoren

**Comment by jcarpent** [comment]
Thanks @Michiel283 for raising this issue. 
I've provided a fix in #2094.

**Comment by Michiel283** [comment]
@jcarpent Thank you so much for the quick response and providing the fix!

**Comment by jcarpent** [comment]
Solved via #2094

**Comment by Michiel283** [comment]
Hi @jcarpent ,

Thank you for releasing a new version of Pinocchio with the CMM correction. I tested the CMM matrices for my system and compared them with my analytical derivation and all matrices were identical. I then computed the centroidal momentum and its time-derivative using the computeCentroidalMomentum(TimeVariation), but the $\dot{\mathbf{h}}_\mathrm{G}$ did not match with my analytic derivation. I think it might be caused by the time-derivative of the transformation to the COM frame again. 
```math
\dot{\mathbf{h}}_\mathrm{G} = {}^\mathrm{G}\mathbf{X}^*_\mathrm{I}\dot{\mathbf{h}}_\mathrm{I} +{}^\mathrm{G}\dot{\mathbf{X}}^*_\mathrm{I}\mathbf{h}_\mathrm{I} \neq {}^\mathrm{G}\mathbf{X}^*_\mathrm{I}\dot{\mathbf{h}}_\mathrm{I}
```
However, the linear part of $\dot{\mathbf{h}}_\mathrm{G}$ is independent on the chain-rule correction above, but this also doesn't match with my analytical derivation. Do you have a suggestion which might explain the discrepancy in the linear centroidal momentum time-variation?
